### PR TITLE
Creating workspaces test suit

### LINKED ISSUE #35: Add a workspace test to prevent incompatible wasm issue happend last week
**Problem**
We hit a incompatible wasm compilation by rustc 1.70 issue. As a result, the incompatible wasm cannot be loaded by nearcore and we cannot call `self_upgrade` to update it. This can only be caught by deploying the contract to a node and test it on CI, before deploy to production.

**Acceptance Criteria** 
- Add a workspace test that deploy the contract and call `self_upgrade` method as sanity check.
- Add it to CI, using the **exactly same** configuration that uses for build and deploy to production

### PR DESCRIPTION
**Pull request by jaswinder6991**
_Resolves #35_

wip PR for creating test suit using workspaces-rs

### CODE REVIEW

**Comment by jaswinder6991** [comment]
Hi @frol, this is the WIP PR for workspaces tests. You can check the migrations issue which occurs if you self_upgrade with the same wasm code that is already deployed. 
I can meanwhile write the test to check wasm incompatibility if any.

**Comment by jaswinder6991** [comment]
Closes https://github.com/near/neardevhub-contract/issues/35

**Comment by frol** [comment]
@jaswinder6991 The issue you observe is related to Wasm binaries produced by Rust 1.70.0 being [not compatible with nearcore runtime](https://github.com/near/nearcore/issues/9143): 

Use the following `rust-toolchain.toml` in the root of the project and re-compile the contract:

```toml
[toolchain]
channel = "1.69.0"
```

The test runs successfully when the contract is compiled with Rust 1.69:

image

**Comment by jaswinder6991** [comment]
Hi @frol, thanks for your comment. I was just testing the case where wasm was not 1.69.0 so sort of a negative case.
Anyway, I was able to deploy 1.71.0 compiled code on testnet. Not sure how?

So for closing https://github.com/near/neardevhub-contract/issues/35, is this PR enough? Should I add some more tests?
I can work on adding this to github workflows. 
@ailisp

**Review by frol — CHANGES_REQUESTED** [review]
Here are the remaining items for this PR to land:

* [x] add CI workflow
* [x] finalize the ergonomics (e.g. `cargo run --example devhub` is clearly not expected way to run tests, try making `cargo test` work)
* [x] check for the result, so it does not rely on human sitting there and reading the output, the test should check and pass if everything is good and fail otherwise
* [x] As part of the test, download the production Wasm from mainnet (https://github.com/near/workspaces-rs#spooning---pulling-existing-state-and-contracts-from-mainnettestnet) and deploy it first, and compile the current version of the contract (https://github.com/near/workspaces-rs#compiling-contracts-during-test-time) as part of the test to deploy it and check migration goes well.

**Comment by jaswinder6991** [comment]
Hi @frol 
The latest commit works on the list of tasks you suggested.
1. CI workflow already has cargo test check, the integration tests execute as a part of cargo test too. Do I need something additional.
2. Tests can be run via "cargo test"
3. I have added assert statements. These are however incomplete at the moment, pending resolving the upgrade issue mentioned in next point.
4. I have successfully used both import_contract and compile_project in the tests. I am currently struggling with the below issue when I call the self_upgrade at [line](https://github.com/jaswinder6991/neardevhub-contract/blob/setup-workspaces/tests/workspaces.rs#L37).
```
status: Failure(ActionError(ActionError { index: Some(0), kind: FunctionCallError(ExecutionError("Exceeded the prepaid gas.")) }
```

Once I am done with the Gas fees issue, I will update the PR. Any help with the issue is updated.

**Comment by jaswinder6991** [comment]
After multiple attempts to solve the below issue (mentioned [here](https://github.com/near/neardevhub-contract/pull/46#issuecomment-1665133158) in more details.).
```
status: Failure(ActionError(ActionError { index: Some(0), kind: FunctionCallError(ExecutionError("Exceeded the prepaid gas.")) }
```
I have written a helper function to mimic compile_project functionality. It does not do error handling as well as the workspaces library but it is working. The reason of above issue produced by workspaces-rs version [0.7.0] is still unknown. However, the workspaces-rs current unreleased codebase do have updates for this function. We can switch to using workspaces-rs if we solve the issue in the future. For noe we continue with the helper function.

File: tests/workspaces.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        .args_json(json!({}))
+        .transact() // note: we use the contract's keys here to sign the transaction
+        .await?;
+
+    assert!(outcome.is_success());
+    assert!(format!("{:?}", outcome).contains("Migrated to version:"));
+
+    println!("Init outcome: {:#?}", outcome);
+
+    //compile the current code
```
Comment (frol) [inline tests/workspaces.rs]: Before moving to the new version, add all types of posts and at least one community, and then after migration, check that get-methods work.

File: tests/README.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+### This subdirectory contains tests for the devhub smart contract. It uses workspaces-rs to create sandbox environment to deploy the contract and run tests.
+
+Use the below command to run the test.
+
+`` cargo test
+``
+
+
```
Comment (frol) [inline tests/README.md]: Can you make this README formatting look nicer?

**Review by frol — CHANGES_REQUESTED** [review]

**Comment by frol** [comment]
I am marking this PR as draft as CI still fails.

File: tests/workspaces.rs (hunk 1 above)
Comment (jaswinder6991) [reply tests/workspaces.rs]: Okay, will work on this.

**Comment by jaswinder6991** [comment]
> I am marking this PR as draft as CI still fails.

The CI worked fine a few weeks ago. Now I am seeing the same Gas Exceeded Error for which I wrote the custom function.
Details mentioned here: https://github.com/near/neardevhub-contract/pull/46#issuecomment-1666830434 

I will look into it again. However, can you think of a reason why self_upgrade would fail with prepaid gas exceeded error.

**Comment by jaswinder6991** [comment]
The Prepaid Gas exceeded issue occured due to lack of gas assigned in self_upgrade function to code deployment and function execution. This PR solves it: https://github.com/near/neardevhub-contract/pull/53

**Review by frol — APPROVED** [review]
@jaswinder6991 I had refactored some bits of the changes, please, review them for your future reference.

**Pull request merged**